### PR TITLE
Improvements to BatchNorm

### PR DESCRIPTION
**Pull request by ShikharJ**
Relevant: #1504 
@zsogitbe Can you confirm if this fixes your issue?

### CODE REVIEW

**Comment by rcurtin** [comment]
Here's a test case we can add:

```
/**
 * Simple serialization test for batch normalization layer.
 */
BOOST_AUTO_TEST_CASE(BatchNormSerializationTest)
{
  arma::mat input(5, 100, arma::fill::randu);
  arma::mat output(5, 100, arma::fill::randu);

  FFN<NegativeLogLikelihood<>, RandomInitialization> model;
  model.Add<Linear<>>(input.n_rows, 10);
  model.Add<BatchNorm<>>();
  model.Add<ReLULayer<>>();
  model.Add<Linear<>>(10, 10);
  model.Add<ReLULayer<>>();
  model.Add<Linear<>>(10, output.n_rows);
  model.Add<LogSoftMax<>>();

  optimization::StandardSGD opt(0.1, 1, 5, -100, false);
  model.Train(input, output, opt);

  arma::mat originalOutput;
  model.Predict(input.col(0), originalOutput);

  // Now serialize the model.
  FFN<NegativeLogLikelihood<>, RandomInitialization> xmlModel, textModel,
      binaryModel;
  SerializeObjectAll(model, xmlModel, textModel, binaryModel);

  // Ensure that predictions are the same.
  arma::mat modelOutput, xmlOutput, textOutput, binaryOutput;
  model.Predict(input.col(0), modelOutput);
  xmlModel.Predict(input.col(0), xmlOutput);
  textModel.Predict(input.col(0), textOutput);
  binaryModel.Predict(input.col(0), binaryOutput);

  CheckMatrices(originalOutput, modelOutput, 1e-5);
  CheckMatrices(originalOutput, xmlOutput, 1e-5);
  CheckMatrices(originalOutput, textOutput, 1e-5);
  CheckMatrices(originalOutput, binaryOutput, 1e-5);
}
```

I think it could be generalized so that we check every layer, but I haven't done it here.  Sure enough in the master branch the test fails:

```
$ bin/mlpack_test -t ANNLayerTest
Running 60 test cases...
unknown location(0): fatal error: in "ANNLayerTest/BatchNormSerializationTest": memory access violation at address: 0x00000000: no mapping at fault address
/home/ryan/src/mlpack/src/mlpack/tests/serialization.hpp(235): last checkpoint

*** 1 failure is detected in the test module "mlpackTest"
```

**Comment by zsogitbe** [comment]
Ryan, I have invested some more time in this issue this Sunday morning and here are my findings about this problem and connected issues:

1. The same problem also exists in LayerNorm.

2. I have tried to serialize 'size' and 'eps' and use weights.set_size when loading the archive, I have also tried to set deterministric = false and reinitialize gamma and beta, but nothing helps. The code still crashes. You will need to dig deeper into this problem. The 'deterministic' parameter (and the stats) may be one of the reasons but I am not completely sure about it because I am still discovering your code.

3. General remark: The compilation of the test cases takes too much time. I think that it would be much better to distribute the tests into separate projects.

4. Your above test, when added to serialization_test.cpp, refuses to compile. It gives an ambitious call to function size() in elu.hpp (line 197 with me), but I guess that you will also see this if you try to compile it.

I hope this helps. Br, Z.

**Comment by zsogitbe** [comment]
Maybe take also a look at running_stat_vec_aux::update_stats(). Would it be possible to serialize a small sample and then construct a running_stat_vec and use the deserialized sample to update the stats?

**Comment by rcurtin** [comment]
We should consider also submitting a patch to Conrad upstream that has a new constructor for `running_stats` that can take in the necessary info (mean, variance, number of samples).  But in the mean time, yeah, replacing `running_stats` with something that's basically identical could work...

I put the test I wrote into `ann_layer_test.cpp` and there were no compilation issues there, though I did also need to `#include "serialization.hpp"` to make the correct symbols available.

**Comment by zoq** [comment]
> 1. The same problem also exists in LayerNorm.

https://github.com/mlpack/mlpack/pull/1508 should fix the the layer norm and batch norm issue; it does pass on the test Ryan suggested above, will test some other cases as well.

> 3. General remark: The compilation of the test cases takes too much time. I think that it would be much better to distribute the tests into separate projects.

You are absolutely right, most of the time I build with ``make -j4`` to use multiple cores for the build. Also, there is an option to discard the tests from the build ``cmake -DBUILD_TESTS=OFF``. Personally, I don't see a benefit of distributing the tests into a separate project, at least not at this point, but if anybody likes the idea and is willing to give it a shot, I'm open for a discussion.

**Comment by zsogitbe** [comment]
Great work Marcus! The generalization of the test is also clever well done!
I confirm that this version passes my own test.
Br, -Z

**Comment by ShikharJ** [comment]
I'll close this pull request as it has been taken up in #1508.
